backend/commons/src/main/java/org/sonarsource/sonarlint/core/commons/util/GitUtils.java

```diff
@@ -0,0 +1,89 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.commons.util;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.eclipse.jgit.ignore.FastIgnoreRule;
+import org.eclipse.jgit.ignore.IgnoreNode;
+import org.eclipse.jgit.lib.Repository;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+
+import static org.eclipse.jgit.lib.Constants.GITIGNORE_FILENAME;
+import static org.sonarsource.sonarlint.core.commons.util.gitblame.GitBlameUtils.buildGitRepository;
+
+public class GitUtils {
+
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+
+  private GitUtils() {
+    // Utility class
+  }
+
+  /**
+   * Assumes the supplied {@param baseDir} or some of its parents is a git repository.
+   * The method checks {@param filePathsToAnalyze} against patterns defined in the .gitignore file and returns
+   * those that are not ignored.
+   * If error occurs during execution then all {@param filePathsToAnalyze} are considered as not ignored.
+   */
+  public static List<URI> filterOutIgnoredFiles(Path baseDir, List<URI> filePathsToAnalyze) {
+    try {
+      var gitRepo = buildGitRepository(baseDir);
+      var gitRepoRelativeProjectBaseDir = gitRepo.getWorkTree().toPath().relativize(baseDir);
+      var ignoreNode = buildIgnoreNode(gitRepo);
+      return ignoreNode.getRules().isEmpty() ? filePathsToAnalyze : filePathsToAnalyze.stream().filter(uri -> {
+        var fileRelativeToGitRepoPath = gitRepoRelativeProjectBaseDir.resolve(Path.of(uri)).toUri().toString();
+        var checkIgnored = checkIgnored(ignoreNode.getRules(), fileRelativeToGitRepoPath);
+        return !checkIgnored.orElse(false);
+      }).collect(Collectors.toList());
+    } catch (Exception e) {
+      LOG.warn("Error occurred while determining files ignored by Git: ", e);
+      LOG.warn("Considering all files as not ignored by Git");
+      return filePathsToAnalyze;
+    }
+  }
+
+  private static IgnoreNode buildIgnoreNode(Repository repository) throws IOException {
+    var rootDir = repository.getDirectory().getParentFile();
+    var gitIgnoreFile = new File(rootDir, GITIGNORE_FILENAME);
+    var ignoreNode = new IgnoreNode();
+    try (var inputStream = new FileInputStream(gitIgnoreFile)) {
+      ignoreNode.parse(inputStream);
+    }
+    return ignoreNode;
+  }
+
+  private static Optional<Boolean> checkIgnored(List<FastIgnoreRule> rules, String entryPath) {
```

Comment:
I would change the return type to the boolean to make it simpler.

Reply:
Fixed in the last commit (34ae30b02d40dedea6c77724771076313aa9ef49)